File: 094_binary_tree_inorder_traversal/Solution2/Solution.go
package main

import "fmt"

type TreeNode struct {
    Val   int
    Left  *TreeNode
    Right *TreeNode
}

func inorderTraversal(root *TreeNode) []int {
    var result []int
    if root == nil { return result }
    var stack []*TreeNode

    for len(stack) != 0 || root != nil {
        if root != nil {
            stack = append(stack, root)
            root = root.Left
        } else {
            lastIdx := len(stack) - 1
            root = stack[lastIdx]
            result = append(result, root.Val)
            stack = stack[:lastIdx]
            root = root.Right
        }
    }

    return result
}

func createTreeUsingBFS(nums []int) *TreeNode {
    arrLen := len(nums)
    var nodeArr []*TreeNode

    for i := 0; i < arrLen; i++ {
        nodeArr = append(nodeArr, &TreeNode{nums[i], nil, nil})
        if i == 0 { continue }
        if i % 2 == 1 {
            nodeArr[(i - 1) / 2].Left = nodeArr[i]
        } else {
            nodeArr[(i - 1) / 2].Right = nodeArr[i]
        }
    }

    return nodeArr[0]
}

func main() {
    nums := []int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13}
    root := createTreeUsingBFS(nums)
    result := inorderTraversal(root)
    for _, num := range result {
        fmt.Printf("%d ", num)
    }
}

File: 094_binary_tree_inorder_traversal/Solution2/Solution.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def create_tree_using_bfs(nums: list) -> TreeNode:
    arr_len = len(nums)
    node_list = []

    for i in range(arr_len):
        new_node = TreeNode(nums[i])
        node_list.append(new_node)
        if i == 0: continue
        if i % 2 == 1: node_list[(i - 1) // 2].left = new_node
        else: node_list[(i - 1) // 2].right = new_node

    return node_list[0]

class Solution:
    def inorderTraversal(self, root: TreeNode) -> list:
        if not root: return []

        stack, result = [], []
        while stack or root:
            if root:
                stack.append(root)
                root = root.left
            else:
                root = stack.pop()
                result.append(root.val)
                root = root.right

        return result


if __name__ == '__main__':
    nums = [i for i in range(14)]
    root = create_tree_using_bfs(nums)
    result = Solution().inorderTraversal(root)
    for num in result:
        print(num, end = ' ')

